libcaf_net/caf/net/datagram_transport.hpp

```diff
@@ -98,17 +98,16 @@ class datagram_transport : public datagram_transport_base<Factory> {
   }
 
   bool handle_write_event(endpoint_manager& manager) override {
-    CAF_LOG_TRACE(CAF_ARG(this->handle_.id)
+    CAF_LOG_TRACE(CAF_ARG2("handle", this->handle_.id)
                   << CAF_ARG2("queue-size", packet_queue_.size()));
-    // Try to write leftover data.
-    write_some();
-    // Get new data from parent.
-    for (auto msg = manager.next_message(); msg != nullptr;
-         msg = manager.next_message()) {
-      this->next_layer_.write_message(*this, std::move(msg));
-    }
-    // Write prepared data.
-    return write_some();
+    auto get_message = [&] {
+      if (auto msg = manager.next_message())
+        this->next_layer_.write_message(*this, std::move(msg));
+      return !packet_queue_.empty();
+    };
+    while (write_some() && get_message())
```

Comment:
Is `&&` really correct? Intuitively, I think we want to try getting the next message if `write_some` failed and stop once we couldn't make any progress, i.e., `write_some() || get_message()`.

Reply:
I think so. `write_some` returns `false` on error which means we don't want to get more data from our parent. If `write_some` returns `true` all queued messages have been sent hence we want to get more messages from our parent. With `||` this would not work that way. We would only stop getting more messages from our parent in the case that there are no more messages to get -- which is not what we want I guess.